ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/parser/TestDirective.java

```diff
@@ -0,0 +1,138 @@
+import com.google.common.base.Functions;
+import io.confluent.ksql.parser.ParsingException;
+import io.confluent.ksql.parser.SqlBaseLexer;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.antlr.v4.runtime.Token;
+
+/**
+ * A test directive indicates special handling of test-only metadata. The
+ * directive is specified in the test sql file as a single-line comment that
+ * begins with * {@code --@} and the directive type and contents are separated
+ * by a colon. The type must contain only letters and "." characters.
+ *
+ * <p>For example, tests are delimited using the {@code TEST} directive, which
+ * would be specified in a file using {@code --@test: name}. The directive types
+ * are case-insensitive.</p>
+ *
+ * <p>The ANTLR grammar that parses comments will handle directives differently
+ * from normal comments, and will send them to the {@link SqlBaseLexer#DIRECTIVES}
+ * channel. See {@link SqlTestReader} for more handling of directives.</p>
+ */
+public class TestDirective {
+
+  static final Pattern DIRECTIVE_REGEX = Pattern.compile(
+      "--@(?<type>[a-zA-Z.]+):\\s*(?<contents>.*)"
+  );
+
+  private final Type type;
+  private final String contents;
+
+  public static TestDirective parse(final Token comment) {
+    final Matcher matcher = DIRECTIVE_REGEX.matcher(comment.getText().trim());
+    if (!matcher.find()) {
+      throw new ParsingException(
+          "Expected directive matching pattern " + DIRECTIVE_REGEX + " but got " + comment,
+          null,
+          comment.getLine(),
+          comment.getCharPositionInLine()
+      );
+    }
+
+    final Type type = Type.from(matcher.group("type").toLowerCase());
+    final String contents = matcher.group("contents");
+
+    return new TestDirective(type, contents);
+  }
+
+  @VisibleForTesting
+  TestDirective(final Type type, final String contents) {
+    this.type = Objects.requireNonNull(type, "type");
+    this.contents = Objects.requireNonNull(contents, "contents");
+  }
+
+  public Type getType() {
+    return type;
+  }
+
+  public String getContents() {
+    return contents;
+  }
+
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    final TestDirective that = (TestDirective) o;
+    return type == that.type
+        && Objects.equals(contents, that.contents);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(type, contents);
+  }
+
+  @Override
+  public String toString() {
+    return "TestDirective{"
+        + "type=" + type
+        + ", contents='" + contents + '\''
+        + '}';
+  }
+
+  public enum Type {
+    TEST("test"),
+    EXPECTED_ERROR("expected.error"),
+    EXPECTED_MESSAGE("expected.message"),
+    UNKNOWN("UNKNOWN")
```

Comment:
What's this for?

If this is just so that `from` can return something, why not make `from` return `Optional<Type>`?

May be worth adding Java docs to each of these types with an explanation of what they do and how they should be used... or better yet... start compiling `.md` files for format of tests.  We have this for QTT. Need it for YATT too.

Reply:
I prefer having an `UNKNOWN` over an optional so that we don't always have to check if present and then just have a switch statement on the enum anyway.

I'll add javadoc, and then when the code is somewhat ready for public consumption I'll add all the formal docs. This push was really to get it in a state where I can use it for query upgrades (and after that this will likely be a main project for me anyway, so I'll have dedicated time on it)

Reply:
Sure, no worries if you feel strongly about it.  

I only raise it as I've seen such patterns before, where the enum includes an 'UNKNOWN' or 'error' value. It can result in the code becoming littered with switch statements that contain the `UNKNOWN` case, which then throw an exception, even though it can never be hit, because some earlier check has already handled it.

Is there any reason why you can't just throw an exception with the location saying 'unknown directive'?

Reply:
My instinct was to have it just ignore unknown directives and move on (as if it were another comment) so that tests can be parsed in previous versions where the directives may not exist. I'm happy also throwing the exception here in parse-time if we don't want that behavior